tests/test_TFNetworkLayer.py

```diff
@@ -150,6 +150,141 @@ def test_concat_sources_missing_dim():
+
+      "att_queries_crop": {"class": "slice", "axis": "t", "slice_start": crop_left, "slice_end": -crop_right, "from": "att_q_split"},
+
+      "att_queries_key_part": {"class": "slice", "axis": "f", "slice_start": 0, "slice_end": querry_dim - context_size, "from": "att_queries_crop"},
+      "att_queries_context_part": {"class": "slice", "axis": "f", "slice_start": querry_dim - context_size, "slice_end": querry_dim, "from": "att_queries_crop"},
+
+
+      "att_key_window": {"class": "window", "window_size": context_size, "padding": "valid", "from": "att_k_split"},
+      "att_energy_dot": {"class": "dot", "red1": "f", "red2": "f", "var1": None, "var2": "s:1", "from": ["att_queries_key_part", "att_key_window"]},
+      "att_energy_scaled": {"class": "eval", "eval": "source(0) * %f" % key_scale, "from": "att_energy_dot"},
+
+
+      "att_energy_biased": {"class": "combine", "kind": "add", "from": ["att_energy_scaled", "att_queries_context_part"]},
+
+      "att_scales": {"class": "activation", "activation": "softmax", "from": "att_energy_biased"},
+
+      "att_value_window": {"class": "window", "window_size": context_size, "padding": "valid", "from": "att_v_split"},
+      "att_out_dot": {"class": "dot", "red1": "f", "red2": "s:1", "var1": None, "var2": "f", "from": ["att_scales", "att_value_window"]},
+
+      "att_out_combine": {"class": "merge_dims", "axes": ["s:1", "f"], "from": "att_out_dot"},
+      "output": {"class": "copy", "from": ["att_out_combine"]}
+    }
+
+    config = Config({"debug_print_layer_output_template": True})
+    in0 = Data(
+      name="in0", shape=(None, n_in // channels, channels), batch_dim_axis=0, auto_create_placeholders=True)
+    extern_data = ExternData()
+    extern_data.register_data(in0)
+
+    network = TFNetwork(config=config, extern_data=extern_data, train_flag=True)
+    network.construct_from_dict(net_dict)
+
+    in_data = rnd.normal(size=(n_batch, n_time, n_in // channels, channels)).astype("float32")
+    in_seq_len = numpy.empty(n_batch)
+    in_seq_len.fill(n_time)
+
+    feed_dict = {in0.placeholder: in_data,
+                 in0.size_placeholder[0]: in_seq_len
+    }
+    fetches = []
+
+    for _, l in network.layers.items():
+      l.output.placeholder = tf.Print(
+          l.output.placeholder, [l.name, "shape:", tf.shape(l.output.placeholder)],
+          summarize=10, name="debug_print_layer_output_shape")
+      fetches.append(l.output.placeholder)
+    output = network.get_default_output_layer().output
+
+    tf.global_variables_initializer().run()
+    out = session.run(fetches=fetches,
+      feed_dict=feed_dict)
+    inp1, inp2 = network.layers["att_energy_dot"].sources[0].output.placeholder, network.layers["att_energy_dot"].sources[1].output.placeholder
+    res = network.layers["att_energy_dot"].output.placeholder
+    inp1, inp2, res = session.run([inp1, inp2, res], feed_dict=feed_dict)
+
+
+def test_CombineLayer():
```

Comment:
Please use more meaningful names. The function name should point out what issue you are testing here (e.g. `test_WindowLayer_ReduceLayer_dim_tags` or so). Later when you understood what the problem is, you can maybe even use a better name. In any case, `CombineLayer` should not be part of the name, as this is likely not the problem (please read again all my comments here in this issue to understand why).

Reply:
Done